consensus/consensus.go

```diff
@@ -1,65 +1,180 @@
+//   - Timeout base duration must be > 0
+//   - The input block must have a valid block hash, block height, view, and QC
+//   - The validators param must be sorted in decreasing order of stake, with a
+//     consistent tie breaking scheme. The validator set is expected to be valid for
+//     validating votes and timeouts for the next block height.
+//
+// Given the above, This function updates the chain tip internally, and initializes all internal
+// data structures that are used to track incoming votes and timeout messages.
+func (fc *FastHotStuffConsensus) Init(
+	blockConstructionCadence time.Duration,
+	timeoutBaseDuration time.Duration,
+	chainTip Block,
+	validators []Validator,
+) error {
+	// Grab the consensus instance's lock
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Ensure the consensus instance is not already running
+	if fc.status == consensusStatusRunning {
+		return errors.New("Consensus instance is already running")
+	}
+
+	// Validate the inputs
+	if blockConstructionCadence <= 0 {
+		return errors.New("Block construction duration must be > 0")
+	}
+	if timeoutBaseDuration <= 0 {
+		return errors.New("Timeout base duration must be > 0")
+	}
+
+	// Update the chain tip and validator set
+	fc.chainTip = chainTip
+	fc.currentView = chainTip.GetView() + 1
+	fc.validators = validators
+
+	// Reset all internal data structures for votes and timeouts
+	fc.votesSeen = make(map[BlockHash]map[bls.PublicKey]VoteMessage)
+	fc.timeoutsSeen = make(map[uint64]map[bls.PublicKey]TimeoutMessage)
+
+	// Reset all internal and external channels used for signaling
+	fc.internalTimersUpdatedSignal = make(chan interface{})
+	fc.stopSignal = make(chan interface{})
+	fc.ConsensusEvents = make(chan *ConsensusEvent)
+
+	// Set the block construction and timeout base durations
+	fc.blockConstructionCadence = blockConstructionCadence
+	fc.timeoutBaseDuration = timeoutBaseDuration
+
+	// Update the consensus status
+	fc.status = consensusStatusInitialized
+
+	return nil
+}
+
+func (fc *FastHotStuffConsensus) UpdateChainTip( /* TODO */ ) {
 	// TODO
 }
 
-func (fc *FastHotStuffConsensus) HandleAcceptedBlock( /* TODO */ ) {
+func (fc *FastHotStuffConsensus) UpdateView( /* TODO */ ) {
 	// TODO
 }
 
-func (fc *FastHotStuffConsensus) HandleVoteMessage( /* TODO */ ) {
+func (fc *FastHotStuffConsensus) ProcessVoteMsg( /* TODO */ ) {
 	// TODO
 }
 
-func (pc *FastHotStuffConsensus) HandleTimeoutMessage( /* TODO */ ) {
+func (pc *FastHotStuffConsensus) ProcessTimeoutMsg( /* TODO */ ) {
 	// TODO
 }
 
-func (fc *FastHotStuffConsensus) HandleBlockProposal( /* TODO */ ) {
+func (fc *FastHotStuffConsensus) ConstructVoteQC( /* TODO */ ) {
 	// TODO
 }
 
+func (fc *FastHotStuffConsensus) ConstructTimeoutQC( /* TODO */ ) {
+	// TODO
+}
+
+// Sets the initial times for the the block construction and timeout timers and starts
+// the event loop building off of the current chain tip.
 func (fc *FastHotStuffConsensus) Start() {
```

Comment:
I think `Start` should kick off the for loop in a go routine. Maybe we should use the same pattern with `startGroup` as in `Mempool`'s Start. This way the caller doesn't have to worry about calling `Start` as `go Start()`. This complexity is pushed inside the `FastHotStuffConsensus` struct. In addition, the `startGroup` will ensure that the goroutine is spawned properly and begun running before we leave the context of `Start`. Lmk if you'd like me to elaborate on how the `startGroup` works.

Moreover, we should update the `fc.status` to `consensusStatusRunning` at the end of `Start`, once `startGroup` for the goroutined loop returns. When we do all of this, we also get a nice simplification of the locks, since we will be able to do a simple lock, defer-unlock.